python/revng/scripting/daemon_project.py

```diff
@@ -0,0 +1,203 @@
+#
+# This file is distributed under the MIT License. See LICENSE.md for details.
+#
+
+import json
+from typing import Any, List
+
+from gql import Client, gql
+from gql.transport.aiohttp import AIOHTTPTransport
+
+from revng.model import Binary, dump_diff_to_string, load_model_from_string, make_diff
+
+from .project import Project
+import yaml
+
+
+class DaemonProject(Project):
+    def __init__(self, url: str, input_binary: str):
+        self.transport = AIOHTTPTransport(url=f"{url}/graphql/")
+        self.client = Client(
+            transport=self.transport, fetch_schema_from_transport=True, execute_timeout=None
+        )
+
+        self.upload_binary(input_binary)
+        self.analyses_list("revng-initial-auto-analysis")
+        self.analyses_list("revng-c-initial-auto-analysis")
+
+        self.set_model(self._get_model())
+        self.pipeline_description = yaml.safe_load(self._get_pipeline_description())
+
+    def upload_binary(self, path: str) -> None:
+        query = gql(
+            """
+            mutation upload($file: Upload!) {
+                uploadFile(file: $file, container: "input")
+            }
+        """
+        )
+
+        with open(path, "rb") as binary_file:
+            result = self.client.execute(query, variable_values={"file": binary_file}, upload_files=True)
+
+        if not result["uploadFile"]:
+            raise RuntimeError("File upload failed")
+
+    def get_artifact(self, artifact_name: str, targets_class: List[Any] = []) -> bytes:
+        self._apply_diff()
+
+        if not targets_class:
+            targets_class = self.model.Functions
+
+        targets = [tc.key() for tc in targets_class]
+
+        query = gql(
+            """
+            query($step: String!, $paths: String!, $index: BigInt!) {
+                produceArtifacts(step: $step, paths: $paths, index: $index) {
+                    ... on Produced {
+                        result
+                    }
+                }
+            }
+        """
+        )
+        variables = {
+            "step": artifact_name,
+            "paths": ",".join(targets),
+            "index": self._get_content_commit_index(),
+        }
+
+        response = self.client.execute(query, variable_values=variables)
+        return response["produceArtifacts"]["result"].encode()
+
+    def get_artifacts(self, params: dict[str, list[Any]]) -> dict:
+        result = {}
+        for artifact_name, targets in params.items():
+            result[artifact_name] = self.get_artifact(artifact_name, targets)
+
+        return result
+
+    def analyze(self, analysis_name: str, target_class: list[Any] = []) -> Binary:
```

Comment:
So I did a few more tests and it looks like you can pass a single function as `target_class` parameter because it causes the `daemon` to crash. Is this expected behavior? Or should `analysis` work on a single `function`?

Reply:
Must have resolved by accident, the crash is beside the point, both points expressed in the first message haven't been addressed

Reply:
In light of recent [events](https://mattermost.rev.ng/revng/pl/7rxtir6wq7bkdc6ara3a6uyw1c), it seems we don't support running `analysis` on `targets`. I removed the `target_class` parameter.

Reply:
This is incorrect, the `analyze` should have a `target_list`. Moreover there should also be a `options` fields for the analysis options.

Reply:
> I understood that passing targets to analyze in the daemon is not supported.

This is wrong, there are 2 aspects:
* In general, analyses need to be able to be provided with a list of targets and options. This is part of the API and it needs to trickle down to the python API
* Specifically, all revng analyses implemented so far do not support being run twice, so when they run they need to run with all the targets, otherwise there would be pieces missing which cannot be computed because again, the analysis cannot run twice